packages/validation-manager/src/ValidationManager.ts

```diff
@@ -82,94 +86,129 @@ export class ValidationManager implements IValidationManager {
     }
   }
 
-  parseValidationResult (userOp: UserOperation, res: ValidationResultStructOutput): ValidationResult {
-    const mergedValidation = mergeValidationDataValues(res.returnInfo.accountValidationData, res.returnInfo.paymasterValidationData)
+  // generate validation result from trace: by decoding inner calls.
+  async generateValidationResult (op: UserOperation, tracerResult: ERC7562Call): Promise<ValidationResult> {
+    // const validationData = tracerResult.calls[0].output
+    let callIndex = 0
+    if (op.factory != null) {
+      callIndex++
+    }
+    if (tracerResult.calls[callIndex] == null) {
+      console.log('fatal: no validateuserop')
+    }
 
-    function fillEntity (addr: string | undefined, info: StakeInfoStructOutput): StakeInfo | undefined {
-      if (addr == null || addr === AddressZero) return undefined
-      return {
-        addr,
-        stake: info.stake,
-        unstakeDelaySec: info.unstakeDelaySec
-      }
+    // pass entrypoint and other addresses we want to resolve by name.
+    const names = {
+      ...op,
+      ep: this.entryPoint.address,
+      senderCreator: await this.entryPoint.senderCreator()
+    }
+    this.erc7562Parser.dumpCallTree(tracerResult, names)
+
+    const validationData = this.decodeValidateUserOp(tracerResult.calls[callIndex])
+    const aggregator = validationData.aggregator
+    let paymasterValidationData: ValidationData = { validAfter: 0, validUntil: maxUint48, aggregator: AddressZero }
+    if (op.paymaster != null) {
+      callIndex++
+      const pmRet = this.decodeValidatePaymasterUserOp(tracerResult.calls[callIndex])
+      // paymasterContext = pmRet.context
+      paymasterValidationData = pmRet.validationData
     }
 
-    const returnInfo = {
-      sigFailed: mergedValidation.aggregator !== AddressZero,
-      validUntil: mergedValidation.validUntil,
-      validAfter: mergedValidation.validAfter,
-      preOpGas: res.returnInfo.preOpGas,
-      prefund: res.returnInfo.prefund
+    // todo: paymasterContext should be returned to parser, to validate its length.
+
+    const paymasterAddress = op.paymaster ?? AddressZero
+    const factoryAddress = op.factory ?? AddressZero
+    const addrs = [op.sender, paymasterAddress, factoryAddress, aggregator]
+    const retStakes = (await runContractScript(this.provider, new GetStakes__factory(), [this.entryPoint.address, addrs]))[0]
```

Comment:
Sorry, what??

Reply:
this is the usage of GetStake. will refactor.